pkg/spanconfig/spanconfigsqlwatcher/sql_watcher.go

```diff
@@ -0,0 +1,453 @@
+func New(
+	codec keys.SQLCodec,
+	DB *kv.DB,
+	settings *cluster.Settings,
+	rangeFeedFactory *rangefeed.Factory,
+	clock *hlc.Clock,
+	ie sqlutil.InternalExecutor,
+	leaseManager *lease.Manager,
+	stopper *stop.Stopper,
+	knobs *spanconfig.TestingKnobs,
+) *SQLWatcher {
+	if knobs == nil {
+		knobs = &spanconfig.TestingKnobs{}
+	}
+	return &SQLWatcher{
+		codec:            codec,
+		DB:               DB,
+		settings:         settings,
+		rangeFeedFactory: rangeFeedFactory,
+		clock:            clock,
+		ie:               ie,
+		leaseManager:     leaseManager,
+		stopper:          stopper,
+		knobs:            knobs,
+	}
+}
+
+// Watch is the entry point for the SQLWatcher.
+func (s *SQLWatcher) Watch(ctx context.Context) (<-chan spanconfig.Update, error) {
+	updatesCh := make(chan spanconfig.Update)
+
+	descUpdatesCh, err := s.watchForDescriptorUpdates(ctx)
+	zonesUpdateCh, err := s.watchForZoneConfigUpdates(ctx)
+	if err != nil {
+		log.Warningf(ctx, "error establishing rangefeed over system.descritpors %v", err)
+		return nil, err
+	}
+	if err := s.stopper.RunAsyncTask(ctx, "span-config-reconciliation", func(ctx context.Context) {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-s.stopper.ShouldQuiesce():
+				return
+			case descID := <-descUpdatesCh:
+				err := s.onDescIDUpdate(ctx, descID, updatesCh)
+				if err != nil {
+					log.Errorf(ctx, "could not react to desc id %d update err: %v", descID, err)
+				}
+			case descID := <-zonesUpdateCh:
+				err := s.onDescIDUpdate(ctx, descID, updatesCh)
+				if err != nil {
+					log.Errorf(ctx, "could not react to zone config update with id %d err: %v", descID, err)
+				}
+			}
+		}
+	}); err != nil {
+		return nil, err
+	}
+	return updatesCh, nil
+}
+
+// watchForDescriptorUpdates establishes a rangefeed over system.descriptors and
+// sends updates on the returned channel. The rangefeed performs an initial scan
+// over the table unless indicated otherwise by the
+// SQLWatcherDisableInitialScan testing knob.
+func (s *SQLWatcher) watchForDescriptorUpdates(ctx context.Context) (<-chan descpb.ID, error) {
+	updatesCh := make(chan descpb.ID)
+	descriptorTableStart := s.codec.TablePrefix(keys.DescriptorTableID)
+	descriptorTableSpan := roachpb.Span{
+		Key:    descriptorTableStart,
+		EndKey: descriptorTableStart.PrefixEnd(),
+	}
+	handleEvent := func(ctx context.Context, ev *roachpb.RangeFeedValue) {
+		value := ev.Value
+		if !ev.Value.IsPresent() {
+			// The descriptor was deleted.
+			value = ev.PrevValue
+		}
+		if !value.IsPresent() {
+			return
+		}
+
+		var descriptor descpb.Descriptor
+		if err := value.GetProto(&descriptor); err != nil {
+			logcrash.ReportOrPanic(
+				ctx,
+				&s.settings.SV,
+				"%s: failed to unmarshal descriptor %v",
+				ev.Key,
+				value,
+			)
+		}
+		if descriptor.Union == nil {
+			return
+		}
+
+		table, database, _, _ := descpb.FromDescriptorWithMVCCTimestamp(&descriptor, ev.Value.Timestamp)
```

Comment:
descriptor.Get{Table,Database}() and descpb.GetDescriptorID() seem more appropriate here.